components/raftstore/src/store/peer.rs

```diff
@@ -477,17 +479,167 @@ pub struct ReadyResult {
+    }
+}
+#[derive(Clone, Debug)]
+pub struct UnsafeRecoveryExecutePlanSharedState(Arc<InvokeClosureOnDrop>);
+impl UnsafeRecoveryExecutePlanSharedState {
+    pub fn new(
+        report_context: UnsafeRecoveryReportContext,
+        router: RaftRouter<impl KvEngine, impl RaftEngine>,
+    ) -> Self {
+        let thread_safe_router = Mutex::new(router);
+        let inner = InvokeClosureOnDrop(Box::new(move || {
+            info!("Unsafe recovery, plan execution finished");
+            let router_ptr = thread_safe_router.lock().unwrap();
+            start_unsafe_recovery_report(&*router_ptr, report_context);
+        }));
+        UnsafeRecoveryExecutePlanSharedState(Arc::new(inner))
+    }
+}
+#[derive(Clone, Debug)]
+pub struct UnsafeRecoveryWaitApplySharedState(Arc<InvokeClosureOnDrop>);
+impl UnsafeRecoveryWaitApplySharedState {
+    pub fn new(
+        report_context: UnsafeRecoveryReportContext,
+        router: RaftRouter<impl KvEngine, impl RaftEngine>,
+    ) -> Self {
+        let thread_safe_router = Mutex::new(router);
+        let inner = InvokeClosureOnDrop(Box::new(move || {
+            info!("Unsafe recovery, wait apply finished");
+            let router_ptr = thread_safe_router.lock().unwrap();
+            let fill_out_report =
+                UnsafeRecoveryFillOutReportSharedState::new(report_context, (*router_ptr).clone());
+            (*router_ptr).broadcast_normal(|| {
+                PeerMsg::SignificantMsg(SignificantMsg::UnsafeRecoveryFillOutReport(
+                    fill_out_report.clone(),
+                ))
+            });
+        }));
+        UnsafeRecoveryWaitApplySharedState(Arc::new(inner))
+    }
+}
+#[derive(Clone, Debug)]
+pub struct UnsafeRecoveryFillOutReportSharedState {
+    _closure: Arc<InvokeClosureOnDrop>,
+    reports: Arc<Mutex<Vec<pdpb::PeerReport>>>,
+}
+impl UnsafeRecoveryFillOutReportSharedState {
+    pub fn new(
+        report_context: UnsafeRecoveryReportContext,
+        router: RaftRouter<impl KvEngine, impl RaftEngine>,
+    ) -> Self {
+        let thread_safe_router = Mutex::new(router);
+        let reports = Arc::new(Mutex::new(vec![]));
+        let reports_clone = reports.clone();
+        let closure = InvokeClosureOnDrop(Box::new(move || {
+            let mut store_report = pdpb::StoreReport::default();
+            {
+                let reports_ptr = reports_clone.lock().unwrap();
+                store_report.set_peer_reports(RepeatedField::from_vec((*reports_ptr).to_vec()));
+            }
+            store_report.set_step(report_context.id);
+            let router_ptr = thread_safe_router.lock().unwrap();
+            if let Err(e) = (*router_ptr).send_control(StoreMsg::UnsafeRecoveryReport(store_report))
+            {
+                error!("Unsafe recovery, fail to schedule reporting"; "err" => ?e);
+            }
+            if report_context.exit_force_leader_after_reporting {
+                (*router_ptr).broadcast_normal(|| {
```

Comment:
(ditto)

Reply:
Mostly the same reason as above, it is hard to know who the force leaders are here, since exiting and entering force leader are separated into 2 different phases, and this is not on the performance critical path.